src/connection/host_lookup.c

```diff
@@ -0,0 +1,541 @@
+    }
+
+  if (((*result) = (struct hostent *) malloc (sizeof (struct hostent))) == NULL)
+    {
+      return NULL;
+    }
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+  memcpy ((void *) *result, (void *) hp_buf, sizeof (struct hostent));
+
+  return 0;
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+  if (hp_buf == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return NULL;
+    }
+
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+
+  return ret;
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+  if (hp_buf == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return -1;
+    }
+
+  memcpy ((void *) ret, (void *) hp_buf, sizeof (struct hostent));
+
+  return 0;
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif /* HAVE_GETHOSTBYNAME_R */
+}
+
+int
+getnameinfo_uhost (struct sockaddr *addr, socklen_t addrlen, char *host, size_t hostlen, char *serv, size_t servlen,
+		   int flags)
+{
+  struct hostent *hp = NULL;
+
+  if (prm_get_bool_value (PRM_ID_USE_USER_HOSTS) == USE_GLIBC_HOSTS)
+    {
+      return getnameinfo (addr, addrlen, host, hostlen, NULL, 0, NI_NOFQDN);
+    }
+  else
+    {
+      if ((hp = host_lookup_internal (NULL, addr, IPADDR_TO_HOSTNAME)) != NULL)
+	{
+	  strncpy (host, hp->h_name, hostlen);
+	  host[CUB_MAXHOSTNAMELEN - 1] = '\0';
+
+	  return 0;
+	}
+
+      return EINVAL;
+    }
+
+}
+
+int
+getaddrinfo_uhost (char *node, char *service, struct addrinfo *hints, struct addrinfo **res)
+{
+  struct hostent *hp = NULL;
+  struct addrinfo results_out;
+  struct sockaddr_in addr_convert;
+  struct in_addr *in_addr_buf = NULL;
+
+  if (prm_get_bool_value (PRM_ID_USE_USER_HOSTS) == USE_GLIBC_HOSTS)
+    {
+      return getaddrinfo (node, service, hints, res);
+    }
+  else
+    {
+      hp = host_lookup_internal (node, NULL, HOSTNAME_TO_IPADDR);
+    }
+
+  if (hp == NULL)
+    {
+//err_print fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_ERROR, -ER_FAILED));
+      return EAI_NODATA;
+    }
+
+  if ((in_addr_buf = (struct in_addr *) malloc (sizeof (struct in_addr))) == NULL)
+    {
+      free (hp);
+//er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct in_addr));
+      return EAI_MEMORY;
+    }
+
+  /*Constitute struct addrinfo for the out parameter res */
+  if (((*res) = (struct addrinfo *) malloc (sizeof (struct addrinfo))) == NULL)
+    {
+      free (hp);
+      free (in_addr_buf);
+//er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (struct in_addr));
+      return EAI_MEMORY;
+    }
+
+  memset (&results_out, 0, sizeof (results_out));
+  if ((results_out.ai_addr = (struct sockaddr *) malloc (sizeof (struct sockaddr))) == NULL)
+    {
+      free (hp);
+      free (in_addr_buf);
+      freeaddrinfo (*res);
```

Comment:
Is it possible to free the memory of res in freeaddrinfo()?
In my opinion it doesn't seem possible. (Same for line 512)
Please refer to the function description below.

The freeaddrinfo() function frees the memory that was allocated for the dynamically allocated linked list res.

Reply:
[ibm docs](https://www.ibm.com/docs/en/zos/2.1.0?topic=functions-freeaddrinfo-free-addrinfo-storage) shows frees one or more addrinfo structures returned by getaddrinfo.
And I tested then I took below result.

> Host: 127.0.0.1
IPv4 address: 127.0.0.1 (127.0.0.1)
0
before freeaddrinfo 2 1 127.0.0.1
Host: 127.0.0.1
after freeaddrinfo0 1

the part of     
struct sockaddr *ai_addr;
    char            *ai_canonname;
    struct addrinfo *ai_next;
    
    are not read.

Reply:
I understood.
But line 512 needs to be checked.

Reply:
Right. It is not needed to be de-allocated all addrinfo.

I will fix it. Thank you!